### Cosine Similarity

In [1]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity


In [2]:
# Sample documents
doc1 = "The sky is blue."
doc2 = "The sun is bright."

In [3]:
# Vectorize the documents
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform([doc1, doc2])

In [8]:
print(tfidf_matrix)

<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 8 stored elements and shape (2, 6)>
  Coords	Values
  (0, 5)	0.40993714596036396
  (0, 3)	0.5761523551647353
  (0, 2)	0.40993714596036396
  (0, 0)	0.5761523551647353
  (1, 5)	0.40993714596036396
  (1, 2)	0.40993714596036396
  (1, 4)	0.5761523551647353
  (1, 1)	0.5761523551647353


In [10]:
import pandas as pd

tfidf_df = pd.DataFrame(tfidf_matrix.toarray(), columns=vectorizer.get_feature_names_out())
print(tfidf_df)

       blue    bright        is       sky       sun       the
0  0.576152  0.000000  0.409937  0.576152  0.000000  0.409937
1  0.000000  0.576152  0.409937  0.000000  0.576152  0.409937


The output above is a DataFrame representing the TF-IDF matrix for the two sample documents:

- Rows: Each row corresponds to a document. Row 0 represents the first document ("The sky is blue."), and row 1 represents the second document ("The sun is bright.").

- Columns: Each column corresponds to a unique word (feature) extracted from the documents. The words are "blue," "bright," "is," "sky," "sun," and "the."

- Values: The values in the DataFrame are the TF-IDF scores for each word in each document. These scores represent the importance of each word in the context of the corresponding document.

For example:

- The value 0.576152 in row 0, column "blue" indicates the TF-IDF score of the word "blue" in the first document. This score is relatively high, reflecting that "blue" is an important and distinctive word in the context of the first document.

- The value 0.000000 in row 0, column "bright" indicates that the word "bright" does not appear in the first document, hence its TF-IDF score is 0.

- The value 0.409937 in row 0, column "is" indicates the TF-IDF score of the word "is" in the first document. This score is lower than that of "blue" because "is" is a more common word and likely carries less distinctive information.

The TF-IDF matrix provides a numerical representation of the documents that can be used for various NLP tasks, such as calculating document similarity, as demonstrated earlier with cosine similarity.

In [4]:
# Calculate cosine similarity
cosine_sim = cosine_similarity(tfidf_matrix[0:1], tfidf_matrix)
print(f"Cosine Similarity: {cosine_sim[0][1]}")

Cosine Similarity: 0.3360969272762575


## Jaccard Similarity

In [11]:
def jaccard_similarity(doc1, doc2):
    words_doc1 = set(doc1.split())
    words_doc2 = set(doc2.split())

    intersection = words_doc1.intersection(words_doc2)

    union = words_doc1.union(words_doc2)

    return len(intersection) / len(union)

In [12]:
# Calculate Jaccard similarity

jaccard_sim = jaccard_similarity(doc1, doc2)
print(f"Jaccard Similarity: {jaccard_sim}")

Jaccard Similarity: 0.3333333333333333


## Euclidean Distance

In [13]:
from sklearn.feature_extraction.text import CountVectorizer
from scipy.spatial.distance import euclidean

In [14]:
# Vectorize the documents

vectorizer = CountVectorizer()
count_matrix = vectorizer.fit_transform([doc1, doc2])

In [18]:
print(count_matrix)

<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 8 stored elements and shape (2, 6)>
  Coords	Values
  (0, 5)	1
  (0, 3)	1
  (0, 2)	1
  (0, 0)	1
  (1, 5)	1
  (1, 2)	1
  (1, 4)	1
  (1, 1)	1


In [19]:
# Convert to dense matrix
dense_matrix = count_matrix.toarray()
print(dense_matrix)

[[1 0 1 1 0 1]
 [0 1 1 0 1 1]]


In [16]:
# Calculate Euclidean distance

euclidean_dist = euclidean(dense_matrix[0], dense_matrix[1])
print(f"Euclidean Distance: {euclidean_dist}")

Euclidean Distance: 2.0
